### PR TITLE
Add lupdate to translation policy

### PR DESCRIPTION
**Pull request by chromatic**

### CODE REVIEW

**Review by patricklodder — COMMENTED** [review]
Concept ACK - thanks!

I was thinking that perhaps we can for now change `make translate` to `lupdate . -ts ./qt/locale/bitcoin_<yourlocale>.ts`, because the latter has a much smaller footprint than the former. It would also allow the `type="vanished"` entries, as the current make macro removes them, which is more useful for Bitcoin's transifex import process than ours.

**Comment by patricklodder** [comment]
_From https://github.com/dogecoin/dogecoin/pull/2869#issuecomment-1060017598_

> I've opened a doc-only PR I've been noodling with to see if it occurs there too.

Doc-only PRs do not trigger CodeQL (and also should not trigger CI, not sure why it does right now - I thought I had fixed that with #2345)

**Comment by chromatic** [comment]
> I was thinking that perhaps we can for now change `make translate` to `lupdate . -ts ./qt/locale/bitcoin_<yourlocale>.ts`, because the latter has a much smaller footprint than the former. It would also allow the `type="vanished"` entries, as the current make macro removes them, which is more useful for Bitcoin's transifex import process than ours.

+1 for dropping `no-obsolete`.

A locale-specific target could be interesting. Do we have a way to get the user's preferred locale from autoconf somehow? I poked around a little bit but nothing jumped out at me. (My `config.status` sets `LC_ALL` and `LANGUAGE` to `C`.)

My only concern is that there may be a tiny niche of people who can work on translations *and* have the Qt dev tools installed. I'm also reluctant to lose a target that allows core devs the ability to update all translations in one, but `make translate-all` is easy enough to define; it's just a copy and paste of the rule as it stands now.

**Comment by chromatic** [comment]
Let's call it good news in general that we *can* reproduce the CodeQL failure here, even if it's still odd that CI ran on this before I touched the build system with the second commit.

**Review by Imshubh69 — DISMISSED** [review]

**Comment by patricklodder** [comment]
blocked by #2882

**Comment by patricklodder** [comment]
@chromatic would you mind rebasing this on the current 1.14.6-dev?

**Comment by chromatic** [comment]
Done!

File: src/Makefile.qt.include
Comment (patricklodder) [inline src/Makefile.qt.include:450]: running `make translate` only updates `bitcoin_en.ts`:

```console
$ make -C src translate 
  GEN      qt/bitcoinstrings.cpp
  GEN      translate
Updating 'qt/locale/bitcoin_en.ts'...
    Found 1073 source text(s) (2 new and 1071 already existing)
    Kept 4 obsolete entries
```

We basically need 2 targets (unlike Bitcoin Core as they use Transifex):

1. Update bitcoin_en.ts (maintainer, standard operation when strings change, simply so that we have a translation file for comparison)
2. Update a particular translation's base strings (contributor, before updating a translation file)

With the below suggestion, this would mean something like this:

```console
$ make -C src translate_base
make: Entering directory '/home/patrick/dogecoin-localbuild/src'
  GEN      translate_base
Updating 'qt/locale/bitcoin_en.ts'...
    Found 1073 source text(s) (2 new and 1071 already existing)
    Removed 4 obsolete entries
make: Leaving directory '/home/patrick/dogecoin-localbuild/src'
$ make -C src translate language=de
make: Entering directory '/home/patrick/dogecoin-localbuild/src'
  GEN      translate
Updating 'qt/locale/bitcoin_de.ts'...
    Found 1073 source text(s) (2 new and 1071 already existing)
    Kept 4 obsolete entries
make: Leaving directory '/home/patrick/dogecoin-localbuild/src'
$ make -C src translate language=en
make: Entering directory '/home/patrick/dogecoin-localbuild/src'
cannot translate English, use make translate_base
make: *** [Makefile:11217: translate] Error 1
make: Leaving directory '/home/patrick/dogecoin-localbuild/src'
$ make -C src translate
make: Entering directory '/home/patrick/dogecoin-localbuild/src'
you need to specify a language
make: *** [Makefile:11216: translate] Error 1
make: Leaving directory '/home/patrick/dogecoin-localbuild/src'
```

```suggestion
translate_base: $(srcdir)/qt/bitcoinstrings.cpp $(QT_FORMS_UI) $(BITCOIN_QT_BASE_CPP) qt/bitcoin.cpp $(BITCOIN_QT_WINDOWS_CPP) $(BITCOIN_QT_WALLET_CPP) $(BITCOIN_QT_H) $(BITCOIN_MM)
  @test -n $(LUPDATE) || echo "lupdate is required for updating translations"
  $(AM_V_GEN) QT_SELECT=$(QT_SELECT) $(LUPDATE) $^ -locations relative -ts $(srcdir)/qt/locale/bitcoin_en.ts
  $(AM_V_GEN) QT_SELECT=$(QT_SELECT) $(LUPDATE) $^ -locations relative -no-obsolete -ts $(srcdir)/qt/locale/bitcoin_en.ts

translate:
  @test -n $(LUPDATE) || echo "lupdate is required for updating translations"
  @if [ "$(language)" = "" ]; then echo "you need to specify a language" && exit 1; fi
  @if [ "$(language)" = "en" ]; then echo "cannot translate English, use make translate_base" && exit 1; fi
  $(AM_V_GEN) QT_SELECT=$(QT_SELECT) $(LUPDATE) $^ -ts $(srcdir)/qt/locale/bitcoin_$(language).ts
```

**Review by patricklodder — CHANGES_REQUESTED** [review]
It doesn't fully work yet as described. I've left a suggestion inline to make it work better for both maintainers and contributors.

File: src/Makefile.qt.include
Comment (chromatic) [reply src/Makefile.qt.include]: Ah, I see what you meant. I've added a change like this and updated the doc to reflect it.

Comment (patricklodder) [reply src/Makefile.qt.include]: Cool, thanks! 

I think that for consistency with how the current translation files work, we should omit `-locations relative` from the actual `translate` target (but keep it for `translate_base`) 

I'd only remove `-no-obsolete` for `translate` as well: It makes no sense to keep it for `translate_base` because there is no manual translation happening there.

Comment (chromatic) [reply src/Makefile.qt.include]: I'm not sure I'm interpreting your `-no-obsolete` suggestion the way you intended, so I've pushed one more commit so that we can discuss this with code.

I'll probably end up squashing that commit into another before it's time to merge.

Comment (patricklodder) [reply src/Makefile.qt.include]: 👍 This is exactly what I meant.

**Comment by patricklodder** [comment]
concept & code review ACK.

re: squash, with a descriptive commit message I think this could be just 1 commit, or 2 if you really want to keep the doc change separate. After that I'll give it a final test and then this would be good to go for me.

**Comment by chromatic** [comment]
Squashed and pushed.

**Review by patricklodder — APPROVED** [review]
tested ACK. Used a Ubuntu 21.04 host with system qt v5.15.2

**Pull request merged**
